src/events/cassandra.py

```
@@ -20,15 +20,21 @@
     RelationDepartedEvent,
     SecretChangedEvent,
     StartEvent,
+    StorageDetachingEvent,
     UpdateStatusEvent,
 )
 from pydantic import ValidationError
-from tenacity import Retrying, stop_after_delay, wait_exponential, wait_fixed
+from tenacity import (
+    Retrying,
+    stop_after_delay,
+    wait_exponential,
+    wait_fixed,
+)
```

**question** why we are not using this? Was this supported to be used on some retry logic in the hook?

---

I don't really know what will be better logic for charm. To retry decommissioning, or to raise after one failed attempt. We could use retrying in future to use locks on multi-unit scaling down.

---

@marceloneppel

---

let's hear @marceloneppel thoughts. If we don't put the retry logic, are we sure that the next hooks are not running?

---

@marcoppenheimer any thoughts too?

---

On PostgreSQL, we used to have some retry logic (and if the charm couldn't finish it after the last retry, we just logged that and let the next event run, because what we needed to do at this moment was only to perform a switchover, which the workload would do later if we can't).

As the decommission process of Cassandra seems to be a different case, where the data needs to be moved to another node, I think it's better to raise an exception like the OpenSearch charm does, to avoid that the unit is removed, so the human operator can call `juju resolve-unit` later to retry the decomission process again after fixing what's causing it not to complete.

---

All right! Understood! I believe this is solid reasoning! Thank you for chiming in!